source/log.c

```diff
@@ -0,0 +1,386 @@
+        aws_raise_error(AWS_ERROR_LOG_UNINITIALIZED);
+        return AWS_OP_ERR;
+    }
+
+    va_list va_args;
+    va_start(va_args, fmt);
+    int ret = aws_vlog(level, fmt, va_args);
+    va_end(va_args);
+    return ret;
+}
+
+const char *aws_log_level_to_string(enum aws_log_level level) {
+    switch (level) {
+        case AWS_LOG_LEVEL_OFF:   return "OFF";
+        case AWS_LOG_LEVEL_FATAL: return "FATAL";
+        case AWS_LOG_LEVEL_ERROR: return "ERROR";
+        case AWS_LOG_LEVEL_WARN:  return "WARN";
+        case AWS_LOG_LEVEL_TRACE: return "TRACE";
+        default: return NULL;
+    }
+}
+
+static void s_aws_log_remove_dead_context(int index) {
+    assert(index >= 0 && index < AWS_LOG_MAX_LOGGING_THREADS);
+
+    /* Atomically grab the ctx pointer for this index. */
+    struct aws_log_context* ctx = aws_atomic_store_ptr((void **)&s_log_table[index].ctx, NULL);
+
+    if (ctx) {
+        /* Cleanup all resources for this context. */
+        struct aws_log_message *msg_list = ctx->message_list;
+        AWS_SINGLY_LIST_REVERSE(struct aws_log_message, msg_list);
+        int count = 0;
+
+        while (msg_list) {
+            struct aws_log_message *next0 = msg_list->next;
+            char *msg_data = (char*)(msg_list + 1);
+            s_log_report_callback(msg_data);
+            aws_memory_pool_release(ctx->message_pool, msg_list);
+            msg_list = next0;
+            ++count;
+        }
+
+        aws_atomic_add(&s_log_message_count, -count);
+
+        struct aws_log_message *delete_list = ctx->delete_list;
+        while (delete_list) {
+            struct aws_log_message *next0 = delete_list->next;
+            aws_memory_pool_release(ctx->message_pool, delete_list);
+            delete_list = next0;
+        }
+
+        void *arena;
+        size_t arena_size;
+        aws_memory_pool_get_arena(ctx->message_pool, &arena, &arena_size);
+        aws_secure_zero(arena, arena_size);
+        aws_memory_pool_clean_up(ctx->message_pool);
+        aws_mem_release(ctx->alloc, ctx);
+    }
+
+    int delete_me = AWS_LOG_ENTRY_STATE_DELETEME;
+    if (aws_atomic_compare_exchange(&s_log_table[index].state, &delete_me, AWS_LOG_ENTRY_STATE_NO_WRITERS)) {
+        aws_atomic_add(&s_log_context_count, -1);
+    }
+}
+
+int aws_log_flush() {
+    for (int i = 0; i < AWS_LOG_MAX_LOGGING_THREADS; ++i) {
+        int state = aws_atomic_store(&s_log_table[i].state, AWS_LOG_ENTRY_STATE_WRITER);
+        if (state != AWS_LOG_ENTRY_STATE_NO_WRITERS) {
+            if (state == AWS_LOG_ENTRY_STATE_DELETEME) {
+                s_aws_log_remove_dead_context(i);
+            }
+            continue;
+        }
+
+        /* Skip empty indices. */
+        struct aws_log_context *ctx = s_log_table[i].ctx;
+        if (!ctx) {
+            s_log_table[i].state = AWS_LOG_ENTRY_STATE_NO_WRITERS;
+            continue;
+        }
+
+        struct aws_log_message *msg_list = aws_atomic_store_ptr((void **)&ctx->message_list, NULL);

```

Comment:
@bdonlan there is a data-dependency on msg_list in this case, so reversing the list can never happen before this line is executed, no?

Reply:
but the next pointers themselves have a data dependency on msg_list, right?

I totally agree with you that it's not clear what memory ordering guarantees this API provides. But just wondering if my understanding is correct.

Isn't it true that regardless of the memory-order semantics used, data-dependency is king? the machine can never re-order those because it would be plain out wrong.

Reply:
I think this came about due to confusion re: the lack of documentation on memory barrier guarantees - if aws_atomic_store_ptr is a full barrier then this works. However, if store_ptr is only ordered wrt that specific location, then we have:

```
CPU 1:
1. READ(message_list)
2. STORE(next)
3. CMPXCHG(message_list)

CPU 2:
4. SWAP(message_list)
5. READ(next)
```

If the operations are not full memory barriers, then from the POV of CPU 2 this could be ordered as follows without violating data dependencies:

```
1. READ(message_list)
2. CMPXCHG(message_list)
4. SWAP(message_list)
5. READ(next)
3. STORE(next)
```